frontend/src/components/UserUpdateRoles.vue

```diff
@@ -1,210 +1,384 @@
 
-      return !!this.$keycloak.tokenParsed.resource_access[umsClientId].roles.includes(manageUserRolesName)
-    }
+      return !!this.$keycloak.tokenParsed.resource_access[
+        umsClientId
+      ].roles.includes(manageUserRolesName);
+    },
   },
   methods: {
-    roleArrayPosition: function(col, item) {
-      return (col - 1) * (this.itemsInColumn) + item - 1;
+    addRoles() {
+      this.selectedClient = null;
+      this.isEdit = false;
+      this.dialog = true;
+    },
+    editRoles: function (client) {
+      this.selectedClient = client;
+      this.isEdit = true;
+      this.dialog = true;
+      this.getUserClientRoles();
     },
-    getClients: function() {
-      return ClientsRepository.get()
-        .then(response => {
-          this.clients = response.data;
+    close() {
+      this.dialog = false;
+    },
+    roleArrayPosition: function (col, item) {
+      return (col - 1) * this.itemsInColumn + item - 1;
+    },
+    addClientRoles: function () {
+      this.isEdit = false;
+      this.getUserClientRoles();
+    },
+    loadUserRoles: function () {
+      this.rolesLoaded = false;
+
+      let results = [];
+      let lastLoginMap = [];
+      let clientsNoRolesAssigned = [];
+
+      UsersRepository.getUserLogins(this.userId).then((lastLogins) => {
+        lastLoginMap = lastLogins.data;
+      });
+
+      ClientsRepository.get()
+        .then((allClients) => {
+          Promise.all(
+            allClients.data.map((client) => {
+              return UsersRepository.getUserEffectiveClientRoles(
+                this.userId,
+                client.id
+              ).then((clientRoles) => {
+                clientRoles["clientRepresentation"] = client;
+                return clientRoles;
+              });
+            })
+          )
+            .then(function (rolesArray) {
+              rolesArray.forEach((clientRoles) => {
+                if (clientRoles.data.length > 0) {
+                  let lastLoginStr = -1;
+                  if (lastLoginMap[clientRoles.clientRepresentation.name]) {
+                  lastLoginStr =
+                    lastLoginMap[clientRoles.clientRepresentation.name];
+                }
+                  clientRoles.roleArray = [];
+                  clientRoles.data.forEach((role) => {
+                    clientRoles.roleArray.push(role.name);
+                  });
+
+                  results.push({
+                    clientRepresentation: clientRoles.clientRepresentation,
+                    roleArray: clientRoles.roleArray,
+                    lastLogin: lastLoginStr,
+                  });
+                } else {
+                  clientsNoRolesAssigned.push(clientRoles.clientRepresentation);
+                }
+              });
+            })
+            .then((this.ClientsWithEffectiveRoles = results))
+            .then((this.clientWithoutRoles = clientsNoRolesAssigned))
+            .finally(() =>this.rolesLoaded = true);
```

Comment:
These don't really need their own .then() blocks, they're not async methods
this could all be

`.finally(() => {
  this.ClientsWithEffectiveRoles = results
  this.clientWithoutRoles = clientsNoRolesAssigned
  this.rolesLoaded = true
})`